chacha20_poly1305/src/chacha20.rs

```diff
@@ -0,0 +1,460 @@
+    fn quarter_round(a: U32x4, b: U32x4, c: U32x4, d: U32x4) -> (U32x4, U32x4, U32x4, U32x4) {
+        let a = a.wrapping_add(b);
+        let d = d.bitxor(a).rotate_left(16);
+
+        let c = c.wrapping_add(d);
+        let b = b.bitxor(c).rotate_left(12);
+
+        let a = a.wrapping_add(b);
+        let d = d.bitxor(a).rotate_left(8);
+
+        let c = c.wrapping_add(d);
+        let b = b.bitxor(c).rotate_left(7);
+
+        (a, b, c, d)
+    }
+
+    // Perform a round on "columns" and then "diagonals" of the state.
+    //
+    // The column quarter rounds are made up of indexes: `[0,4,8,12]`, `[1,5,9,13]`, `[2,6,10,14]`, `[3,7,11,15]`.
+    // The diagonals quarter rounds are made up of indexes: `[0,5,10,15]`, `[1,6,11,12]`, `[2,7,8,13]`, `[3,4,9,14]`.
+    //
+    // The underlying quarter_round function is vectorized using the
+    // u32x4 type in order to perform 4 quarter round functions at the same time.
+    // This is a little more difficult to read, but it gives the compiler
+    // a strong hint to use the performant SIMD instructions.
+    #[inline(always)]
+    fn double_round(a: U32x4, b: U32x4, c: U32x4, d: U32x4) -> (U32x4, U32x4, U32x4, U32x4) {
+        // Vectorized column quarter round.
+        let (a, b, c, d) = Self::quarter_round(a, b, c, d);
+
+        // Rotate the words into their diagonal positions
+        // in order to efficiently use the same quarter round function.
+        let b = b.rotate_elements_left::<1>();
+        let c = c.rotate_elements_left::<2>();
+        let d = d.rotate_elements_left::<3>();
+
+        // Vectorized diagonal quarter round.
+        let (a, b, c, d) = Self::quarter_round(a, b, c, d);
+
+        // Rotate the words back into their normal positions.
+        (
+            a,
+            b.rotate_elements_right::<1>(),
+            c.rotate_elements_right::<2>(),
+            d.rotate_elements_right::<3>(),
+        )
+    }
+
+    /// Transform the state by performing the ChaCha block function.
+    fn chacha_block(&mut self) {
+        // State is pulled out into temporary variables
+        // to make the function more vectorizable.
+        let mut a = self.a;
+        let mut b = self.b;
+        let mut c = self.c;
+        let mut d = self.d;
+
+        for _ in 0..10 {
+            (a, b, c, d) = Self::double_round(a, b, c, d);
+        }
+
+        // Add the initial state to the final state
+        self.a = a.wrapping_add(self.a);
+        self.b = b.wrapping_add(self.b);
+        self.c = c.wrapping_add(self.c);
+        self.d = d.wrapping_add(self.d);
+    }
+
+    /// Expose the 512-bit state as a byte stream.
+    fn keystream(&self) -> [u8; 64] {
+        let mut keystream = [0u8; 64];
+
+        keystream[0..16].copy_from_slice(&self.a.to_le_bytes());
+        keystream[16..32].copy_from_slice(&self.b.to_le_bytes());
+        keystream[32..48].copy_from_slice(&self.c.to_le_bytes());
+        keystream[48..64].copy_from_slice(&self.d.to_le_bytes());
```

Comment:
Wouldn't it be better to store `a`, `b`, `c`, `d` as an array and then use a loop here?

Reply:
This was a SIMD influenced decision. Let me run some tests to be sure it is needed, but in any case, I'll leave a comment since it is not immediately obvious.